src/vs-picgo/index.ts

```diff
@@ -0,0 +1,192 @@
+import * as fs from 'fs';
+import * as path from 'path';
+import * as os from 'os';
+import * as vscode from 'vscode';
+
+import PicGo from 'picgo/dist/core/PicGo';
+import { ImgInfo } from 'picgo/dist/utils/interfaces';
+
+import { promisify } from 'util';
+
+import { formatParam, formatString } from '../utils';
+
+const _ = require('lodash');
+const _db = require('lodash-id');
+_.mixin(_db);
+
+const writeFileP = promisify(fs.writeFile);
+const readFileP = promisify(fs.readFile);
+
+export interface INotice {
+  body: string;
+  text: string;
+  title: string;
+}
+
+export interface IUploadName {
+  date: string;
+  dateTime: string;
+  fileName: string;
+  extName: string;
+  mdFileName: string;
+  [key: string]: string;
+}
+
+export interface IOutputUrl {
+  uploadedName: string;
+  url: string;
+  [key: string]: string;
+}
+
+export default class VSPicgo {
+  private dataPath: string;
+  private editor: vscode.TextEditor;
+  private picgo: PicGo;
+
+  constructor(editor: vscode.TextEditor) {
+    const picgoConfigPath = vscode.workspace.getConfiguration('picgo').get<string>('configPath');
+    if (picgoConfigPath) {
+      this.picgo = new PicGo(picgoConfigPath);
+    } else {
+      this.picgo = new PicGo();
+      const picBed = vscode.workspace.getConfiguration('picgo.picBed');
+      this.picgo.setConfig({ picBed });
+    }
+    this.editor = editor;
+    this.dataPath = this.getDataPath();
+    // Before upload, we change names of the images.
+    this.addRenameListener();
+    // After upload, we use the custom output format.
+    this.addGenerateOutputListener();
+  }
+
+  addGenerateOutputListener() {
+    this.picgo.on('finished', async (ctx: PicGo) => {
+      let urlText = '';
+      const outputFormatTemplate =
+        vscode.workspace.getConfiguration('picgo').get<string>('customOutputFormat') || '![${uploadedName}](${url})';
+      try {
+        urlText = ctx.output.reduce((acc: string, cur: ImgInfo): string => {
+          return `${acc}${formatString(outputFormatTemplate, {
+            uploadedName: cur.fileName || '',
+            url: cur.imgUrl,
+          })}\n`;
+        }, '');
+        urlText = urlText.trim();
+        await this.updateLog(ctx.output);
```

Comment:
By the way, I think these `log` & `Log` should be changed to `data` and `Data` now to avoid confusion.

Reply:
Yep!